### ISSUE TITLE
How do I shim a module from the window?

### ISSUE THREAD

**Issue by rohmann** [ef4/ember-auto-import#200]
I'm in a unique situation where I need to setup an import that simply resolves back to an existing reference on the window. For example, in my app I have:

```
import _ from 'lodash'
```

and I get this error:

```
Uncaught Error: Could not find module `lodash` imported from `tco-app/components/react-component`
```

I've tried this configuration:

```
autoImport: {
      alias: {
        'lodash': './shims/lodash.js',
    }
}
```

Where `app/shims/lodash.js` contains: `module.exports = window.lodash`

Similarly, I've tried adding `browser` to `package.json`
```
"browser": {
    "lodash": "./shims/lodash.js"
}
```

I don't have `lodash` in my dependencies which causes the analyzer to ignore it. Also, the same thing happens if I explicitly exclude the package via `autoImport` config. I've tried using webpack externals but couldn't get past the error above.

Oddly enough, I can get other packages that depend on `lodash` to use the version defined on `window.lodash` by adding this:

```
autoImport: {
  webpack: {
    resolve: {
      alias: {
        'lodash': path.resolve(__dirname, 'shims/lodash' )
      }
    }
  }
}
```

Unfortunately that only works for other packages and not my own files.

TL;DR; So what's the best way to make `import _ from 'lodash'` always resolve to the existing `window.lodash`? Thanks!

**Comment by rohmann** [comment]
I've poked around a bit more and seems like something that may just not be supported.

This is obviously a hack but by checking if `aliasedSpecifier` is a local file (e.g. `./shims/lodash.js`) I can bypass the module not being recognized

https://github.com/ef4/ember-auto-import/blob/master/packages/ember-auto-import/ts/splitter.ts#L89-L99

```
if (0 !== aliasedSpecifier.indexOf('./')) {
    if (imp.package.excludesDependency(packageName)) {
      // This package has been explicitly excluded.
      return;
    }

    if (
      !imp.package.hasDependency(packageName) ||
      imp.package.isEmberAddonDependency(packageName)
    ) {
      return;
    }
}
```

Could I request an official (not a hack) solution for this? Thanks you!

**Comment by ef4** [comment]
If you already have `window.lodash` and you want to make it importable, make:

```js
// vendor/shims/lodash.js
(function() {
  function vendorModule() {
    'use strict';

    return {
      'default': self.lodash,
      __esModule: true,
    };
  }

  define('lodash', [], vendorModule);
})();
```

And call `app.import('vendor/shims/lodash.js')` in your ember-cli-build.js file.

That is the non-ember-auto-import way of doing it. Alternatively if you want to use ember-auto-import, remove whatever is currently creating `window.lodash`, add lodash as a true `dependency`, and auto import it. But what you're asking above sounds like "how do I use ember-auto-import without using ember-auto-import?".

A different scenario would be if there is some third-party, auto-imported code that wants to import lodash. In that case, `externals` is indeed the right solution. The difference in that case is that the library doing the importing is a library that was auto-imported, which means its getting bundled by webpack and wants to access a thing that's not bundled by webpack.

In your case you have app code (which is not bundled by webpack) that wants to access a global (which is not bundled by webpack), so ember-auto-import is not really involved at all.

**Comment by rohmann** [comment]
Thank you! That's exactly what I needed. Looks like I took the long way around on this one... I really appreciate the guidance on getting that shim in place.
